docs/architecture/adr-064-abci-2.0.md

```diff
@@ -0,0 +1,443 @@
+			Round:     int64(extCommit.Round),
+			ChainId:   app.GetChainID(),
+		}
+
+		extSignBytes, err := cosmosio.MarshalDelimited(&cve)
+		if err != nil {
+			return fmt.Errorf("failed to encode CanonicalVoteExtension: %w", err)
+		}
+
+		if !cmtPubKey.VerifySignature(extSignBytes, vote.ExtensionSignature) {
+			return errors.New("received vote with invalid signature")
+		}
+
+		return nil
+	}
+}
+```
+
+Once at least 2/3 signatures, by voting power, are received and verified, the
+validator can use the vote extensions to derive additional data or come to some
+decision based on the vote extensions.
+
+> NOTE: It is very important to state, that neither the vote propagation technique
+> nor the vote extension verification mechanism described above is required for
+> applications to implement. In other words, a proposer is not required to verify
+> and propagate vote extensions along with their signatures nor are proposers
+> required to verify those signatures. An application can implement it's own
+> PKI mechanism and use that to sign and verify vote extensions.
+
+#### Vote Extension Persistence
+
+In order to make any data derived from vote extensions persistent, we propose to
+to allow application developers to "merge" the `processProposalState` into the
+`finalizeState`, such that when processing transactions during `FinalizeBlock`,
+the base state includes any state changes from processing votes extensions
+(see [`FinalizeBlock`](#finalizeblock-1) below).
+
+However, we DO NOT want to pollute `finalizeState` with any state changes from
+verifying transactions during `ProcessProposal`, so we will introduce a new API,
+`ResetProcessProposalState`, that will allow applications to essentially reset
+the `ProcessProposal` state, effectively allowing any state transitions after to
+be committed.
+
+What this means is that any explicit state changes in a `ProcessProposal` handler
+using the injected `processProposalState.Context` will be written to state! An
+application must be careful to execute `ResetProcessProposalState` where and when
+appropriate.
+
+A `ProcessProposal` handler could look like the following:
+
+```go
+func (h MyHandler) ProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req abci.RequestProcessProposal) abci.ResponseProcessProposal {
+		for _, txBytes := range req.Txs {
+			_, err := h.app.ProcessProposalVerifyTx(txBytes)
+			if err != nil {
+				return abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}
+			}
+		}
+
+		// Reset the ProposalProposal state, which, up to this point, has accumulated
+		// possible state transitions from transaction verification and processing.
+		// Any subsequent state transitions after this will be merged into the
+		// FinalizeState before processing Begin/EndBlock and transactions.
+		// 
+		// NOTE: Applications must call this to ensure no state transactions up to now
+		// are committed!
+		h.app.ResetProcessProposalState()
```

Comment:
What's the reasoning for adding `ResetProcessProposalState`? 

Is it because we want to preserve the work done in `ProcessProposal` to `FinalizeBlock` in certain use cases?